File: Engine/src/main/java/org/etieskrill/engine/entity/component/Drawable.java
package org.etieskrill.engine.entity.component;

import org.etieskrill.engine.graphics.gl.shader.ShaderProgram;
import org.etieskrill.engine.graphics.model.Model;
import org.jetbrains.annotations.Nullable;
import org.joml.Vector2f;

public class Drawable {

    private final Model model;
    private boolean visible;
    private final Vector2f textureScale;
    private boolean drawWireframe;
    private @Nullable ShaderProgram shader;
    private boolean drawOutline;
    private float outlineWidth;

    //TODO some kind of grouping mechanism

    public Drawable(Model model) {
        this(model, null);
    }

    public Drawable(Model model, @Nullable ShaderProgram shader) {
        this.model = model;
        this.visible = true;
        this.textureScale = new Vector2f(1);
        this.drawWireframe = false;
        this.shader = shader;
        this.drawOutline = false;
        this.outlineWidth = 0.05f;
    }

    public Model getModel() {
        return model;
    }

    public boolean isVisible() {
        return visible;
    }

    public void setVisible(boolean visible) {
        this.visible = visible;
    }

    public Vector2f getTextureScale() {
        return textureScale;
    }

    public void setTextureScale(Vector2f textureScale) {
        this.textureScale.set(textureScale);
    }

    public boolean isDrawWireframe() {
        return drawWireframe;
    }

    public void setDrawWireframe(boolean drawWireframe) {
        this.drawWireframe = drawWireframe;
    }

    public @Nullable ShaderProgram getShader() {
        return shader;
    }

    public void setShader(@Nullable ShaderProgram shader) {
        this.shader = shader;
    }

    public boolean isDrawOutline() {
        return drawOutline;
    }

    public void setDrawOutline(boolean drawOutline) {
        this.drawOutline = drawOutline;
    }

    public float getOutlineWidth() {
        return outlineWidth;
    }

    public void setOutlineWidth(float outlineWidth) {
        this.outlineWidth = outlineWidth;
    }

}

File: Engine/src/jvmMain/java/org/etieskrill/engine/entity/component/Drawable.kt
package org.etieskrill.engine.entity.component

import org.etieskrill.engine.common.Disposable
import org.etieskrill.engine.graphics.gl.shader.ShaderProgram
import org.etieskrill.engine.graphics.model.Model
import org.joml.Vector2f
import org.joml.Vector2fc

class Drawable(
    val model: Model,
    var shader: ShaderProgram? = null,
    var isVisible: Boolean = true,
    var isWireframeEnabled: Boolean = false,
    var isOutlineEnabled: Boolean = false,
    var outlineWidth: Float = 0.05f,
    var textureScale: Vector2fc? = Vector2f(1f),
) : Disposable {

    override fun dispose() {
        model.dispose() //TODO somehow mark owning and non-owning members
        shader?.dispose()
    }

}
